### PR TITLE
Convert group work info container to component

### PR DESCRIPTION
**Pull request by jonatanschroeder**

### CODE REVIEW

File: apps/prairielearn/src/tests/groupRole.test.ts [hunk 1]
```diff
@@ -1516,14 +1475,11 @@ describe('Test group role reassignments with role of minimum > 1', function () {
...
     step('should have correct errors displayed', function () {
-      elemList = locals.$(
-        '.alert:contains(1 more student needs to be assigned to the role "Recorder" (Found 1, expected exactly 2).)',
-      );
-      assert.lengthOf(elemList, 0);
-      elemList = locals.$(
-        '.alert:contains(At least one user does not have a role. All users must have a role.)',
+      assertAlert(
+        locals.$,
+        '1 more student needs to be assigned to the role "Recorder" (1 assigned, 2 expected)',
```
Comment (jonatanschroeder) [inline apps/prairielearn/src/tests/groupRole.test.ts:1480]: This seems to be a case where a recent refactor fixed this test the wrong way. The error should indeed display, but the text was incorrect.

File: apps/prairielearn/src/components/GroupWorkInfoContainer.html.ts [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+    <div class="container-fluid">
+      <div class="row">
+        <div class="col-sm bg-light py-4 px-4 border">
+          <h6 style="display:inline"><b>Group name: </b></h6>
+          <p style="display:inline" id="group-name">${groupInfo.groupName}</p>
+          <br />
+          ${groupConfig.student_authz_join
+            ? html`
+                <h6 style="display:inline"><b>Join code: </b></h6>
+                <p style="display:inline" id="join-code">${groupInfo.joinCode}</p>
```
Comment (jonatanschroeder) [inline apps/prairielearn/src/components/GroupWorkInfoContainer.html.ts:30]: Using `h6` and `b` here, as well as a `p` with inline display, bugs me a bit. Any suggestion for something a bit more semantic?

File: apps/prairielearn/src/tests/helperClient.ts [hunk 3]
```diff
@@ -200,3 +200,18 @@ export async function saveOrGrade(
     }).toString(),
   });
 }
+
+/**
+ * Asserts that an alert exists with a given text. Normalizes the text of the
+ * alert before comparing with the expected value.
+ */
+export function assertAlert($: cheerio.CheerioAPI, text: string, expectedLength = 1) {
```
Comment (jonatanschroeder) [inline apps/prairielearn/src/tests/helperClient.ts:208]: This was created because Cheerio's `:contains` does not normalize the text of the alert. Since the component includes some indentation in the text, the text will not match exactly because the spaces don't match. This function then normalizes the text of the alert before comparing the content. While doing so, it also prints the existing alerts if the alert in question does not exist, making debugging easier.

File: apps/prairielearn/src/components/GroupWorkInfoContainer.html.ts (hunk 2 above)
Comment (nwalters512) [reply apps/prairielearn/src/components/GroupWorkInfoContainer.html.ts:30]: Agree that this is not great. I think `<p><strong>Join code:</strong> ...</p>` would be fine.

File: apps/prairielearn/src/components/GroupWorkInfoContainer.html.ts [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+                          : ''}
+                      `}
+                </div>
+              `
+            : ''}
+        </div>
+        <div class="col-sm bg-light py-4 px-4 border">
+          ${groupConfig.student_authz_leave
+            ? html`
+                <div style="display: flex; justify-content: flex-end">
```
Comment (nwalters512) [inline apps/prairielearn/src/components/GroupWorkInfoContainer.html.ts:57]: Nit: I imagine this was just copy/paste, but can we use Bootstrap classes (`d-flex`, etc.) instead?

File: apps/prairielearn/src/components/GroupWorkInfoContainer.html.ts [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+                  <button
+                    type="button"
+                    class="btn btn-danger"
+                    data-toggle="modal"
+                    data-target="#leaveModal"
```
Comment (nwalters512) [inline apps/prairielearn/src/components/GroupWorkInfoContainer.html.ts:62]: Nit: let's call this `leaveGroupModal`?

File: apps/prairielearn/src/components/GroupWorkInfoContainer.html.ts [hunk 6]
```diff
@@ -0,0 +1,283 @@
+import { html } from '@prairielearn/html';
+
+import type { GroupConfig } from '../lib/db-types.js';
+import type { GroupInfo } from '../lib/groups.js';
+import { idsEqual } from '../lib/id.js';
+
+import { Modal } from './Modal.html.js';
+
+export function GroupWorkInfoCountainer({
```
Comment (nwalters512) [inline apps/prairielearn/src/components/GroupWorkInfoContainer.html.ts:9]:
```suggestion
export function GroupWorkInfoContainer({
```

File: apps/prairielearn/src/components/GroupWorkInfoContainer.html.ts [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+              : 'ask a user with an assigner role to update the role configuration'}.
+            Question submissions are disabled until the role configuration is valid.
+          </div>
+        `
+      : ''}
+    <details class="card mb-2">
+      <summary class="card-header bg-secondary text-light">
+        ${userCanAssignRoles ? 'Manage group roles' : 'View group roles'}
+        ${roleConfigProblems > 0
+          ? html` <span class="badge badge-pill badge-danger">${roleConfigProblems}</span> `
```
Comment (nwalters512) [inline apps/prairielearn/src/components/GroupWorkInfoContainer.html.ts:142]: Nit:

```suggestion
          ? html`<span class="badge badge-pill badge-danger">${roleConfigProblems}</span>`
```

File: apps/prairielearn/src/pages/studentAssessment/studentAssessment.js [hunk 8]
```diff
@@ -83,32 +83,22 @@ router.get(
...
+      const userCanAssignRoles =
+        groupInfo != null &&
+        groupConfig.has_roles &&
+        (canUserAssignGroupRoles(groupInfo, res.locals.user.user_id) ||
+          res.locals.authz_data.has_course_instance_permission_edit);
+      res.send(
+        StudentAssessment({ resLocals: res.locals, groupConfig, groupInfo, userCanAssignRoles }),
+      );
+    } else {
+      res.send(StudentAssessment({ resLocals: res.locals }));
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/studentAssessment/studentAssessment.js:100]: Nit: instead check `!res.locals.assessment_group_work` and early-return so we can reduce indentation of the above?

File: apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.js [hunk 9]
```diff
@@ -291,14 +285,24 @@ router.get(
...
+        StudentAssessmentInstance({
+          showTimeLimitExpiredModal: req.query.timeLimitExpired === 'true',
+          resLocals: res.locals,
+          groupConfig,
+          groupInfo,
+          userCanAssignRoles,
+        }),
+      );
+    } else {
+      res.send(
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.js:299]: Nit, same comment as above, maybe early-return intead.

File: apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.js [hunk 10]
```diff
@@ -257,24 +257,18 @@ router.get(
...
-
-      res.locals.notInGroup = false;
       const groupInfo = await getGroupInfo(res.locals.assessment_instance.group_id, groupConfig);
-      res.locals.groupSize = groupInfo.groupSize;
-      res.locals.groupMembers = groupInfo.groupMembers;
-      res.locals.joinCode = groupInfo.joinCode;
-      res.locals.groupName = groupInfo.groupName;
-      res.locals.start = groupInfo.start;
-      res.locals.rolesInfo = groupInfo.rolesInfo;
-      res.locals.used_join_code = req.body.used_join_code;
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/studentAssessmentInstance/studentAssessmentInstance.js:273]: 🎉

File: apps/prairielearn/src/tests/groupExam.test.ts [hunk 11]
```diff
@@ -66,7 +67,7 @@ async function switchUserAndLoadAssessment(
   const $ = cheerio.load(page);
 
   // Check that the correct CSRF form exists
-  const elementQuery = `form[name="${formName}"] input[name="__csrf_token"]`;
+  const elementQuery = `form${formName ? `[name="${formName}"]` : ''} input[name="__csrf_token"]`;
```
Comment (nwalters512) [inline apps/prairielearn/src/tests/groupExam.test.ts:70]: Why this change? When would we want to grab *any* CSRF token on the page? This makes this test less useful as it doesn't actually assert that the form we're submitting has a CSRF token in it.

File: apps/prairielearn/src/components/GroupWorkInfoContainer.html.ts (hunk 5 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/components/GroupWorkInfoContainer.html.ts]: Done in abcb6d8d3.

File: apps/prairielearn/src/tests/groupExam.test.ts (hunk 11 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/tests/groupExam.test.ts]: This change was necessary because the leave group form is now using a modal form, which doesn't have a name. I have updated it in abcb6d8d3 to use a container instead in this case, so that instead of finding a form from the name, you find the form inside the modal.

File: apps/prairielearn/src/components/GroupWorkInfoContainer.html.ts (hunk 4 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/components/GroupWorkInfoContainer.html.ts]: Done in 01ad7596b.

File: apps/prairielearn/src/components/GroupWorkInfoContainer.html.ts (hunk 2 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/components/GroupWorkInfoContainer.html.ts]: Done in 3340238d6.

File: apps/prairielearn/src/components/GroupWorkInfoContainer.html.ts (hunk 6 above)
Comment (jonatanschroeder) [reply apps/prairielearn/src/components/GroupWorkInfoContainer.html.ts]: 🤦🏻‍♂️  3340238d6

**Comment by nwalters512** [comment]
@jonatanschroeder I saw the review request, there are a lot of tests failing though.

**Comment by jonatanschroeder** [comment]
> @jonatanschroeder I saw the review request, there are a lot of tests failing though.

I had tested locally, but the error was something that didn't show up in the UI. In particular, the issue was that it was missing the `return;` after the early send, but since the send had already happened the error did not show up on the page, only on the terminal. This was picked up by CI, though. Sorry about that.

**Review by nwalters512 — APPROVED** [review]

**Comment by nwalters512** [comment]
No worries @jonatanschroeder, just wanted to make sure you were aware of the failures!
